src/kernels/gibbskernel.jl

```diff
@@ -0,0 +1,44 @@
+"""
+    GibbsKernel(x, y)
+
+# Definition
+
+The Gibbs kernel is non-stationary generalisation of the Squared-Exponential
+kernel. The lengthscale parameter ``l`` becomes a function of
+position ``l(x)``.
+
+``l(x) = 1.`` then you recover the standard Squared-Exponential kernel
+with constant lengthscale.
+
+```math
+k(x, y) = \\sqrt{ \\left(\\frac{2 l(x) l(y)}{l(x)^2 + l(y)^2} \\right) }
+\\quad \\rm{exp} \\left( - \\frac{(x - y)^2}{l(x)^2 + l(y)^2} \\right)
+```
+
+[1] - Mark N. Gibbs. "Bayesian Gaussian Processes for Regression and Classication."
+    PhD thesis, 1997
+
+[2] - Christopher J. Paciorek and Mark J. Schervish. "Nonstationary Covariance
+    Functions for Gaussian Process Regression". NEURIPS, 2003
+
+[3] - Sami Remes, Markus Heinonen, Samuel Kaski.
+    "Non-Stationary Spectral Kernels". arXiV:1705.08736, 2017
+
+[4] - Sami Remes, Markus Heinonen, Samuel Kaski.
+    "Neural Non-Stationary Spectral Kernel". arXiv:1811.10978, 2018
+"""
+
+struct GibbsKernel{T} <: Kernel
+    lengthscale::T
+end
+
+GibbsKernel(; lengthscale) = GibbsKernel(lengthscale)
+
+function (k::GibbsKernel)(x, y)
+    lengthscale = k.lengthscale
+    lx = lengthscale(x)
+    ly = lengthscale(y)
+    l = hypot(lx, ly)
+    kernel = (sqrt(2 * lx * ly) / l) * with_lengthscale(SqExponentialKernel(), l)
```

Comment:
```suggestion
    l = hypot(lx, ly) / sqrt(2)
    kernel = (sqrt(lx * ly) / l) * with_lengthscale(SqExponentialKernel(), l)
```
I think this should more correspond to the exact formula

Reply:
Without promotion to Float64:
```suggestion
    l = invsqrt2 * hypot(lx, ly)
    kernel = (sqrt(lx * ly) / l) * with_lengthscale(SqExponentialKernel(), l)
```

Reply:
Is this correct though because `l` is passed into `with_lengthscale` - doesn't the `SqExponentialKernel ` already have the `1/2` in the exponential looking [here](https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/blob/master/src/basekernels/exponential.jl#L23)

Reply:
shouldn't it be

```
l = hypot(lx, ly)
kernel = (sqrt(2 * lx * ly) / l) * with_lengthscale(SqExponentialKernel(), 1 / l^2)
```

Reply:
> Is this correct though because l is passed into with_lengthscale - doesn't the SqExponentialKernel already have the 1/2 in the exponential looking here

Yes but in the latex formula the factor 2 is not there, so this way we cancel it out!

Reply:
@devmotion proposition is definitely the right one!

Reply:
oh yeh.... thanks!

Reply:
In general, the factor 2 in the exponential does not matter but it is a bit closer to the formulas in the issue 🤷‍♂️

Reply:
I guess the 2 disappears because we already have l(x) + l(y) ?

Reply:
Yeah the 2 would be there when l(x) = l(y) = constant.

I think I finally understand most of the things here but I am confused by `with_lengthscale`. Can you explain why we shouldn't square the argument to with_lengthscale? Looking at the with_lengthscale code it seems to just multiply by `1/l`.

Reply:
> I guess the 2 disappears because we already have l(x) + l(y) ?

I'm not sure what 2 you refer to (and there is no l(x) + l(y) term anywhere 🤔). The 2 in the square root outside of the squared exponential kernel disappears because we already scale the square root of the denominator by 1/sqrt(2).

> why we shouldn't square the argument to `with_lengthscale`

It will be squared by the squared exponential kernel. A squared exponential kernel with lengthscale l is of the form `k(x, y) = exp(- (x - y)^2/(2*l^2))`.

Reply:
So the `l(x) + l(y)` term I was refering to is the denominator in the exponential i.e.
`exp(-(x - y)^2 / (l(x)^2 + l(y)^2)`

And when `l(x) = l(y) = l = constant` you get `exp(-(x - y)^2 / (2l^2)` ?

Also thanks for explaining to me about the `with_lengthscale`

Reply:
Exactly, if they are constant you get `exp(-(x-y)^2 / (2 * l^2))`. Very explicitly, in general the implementation yields
```math
k(x, y) = sqrt(l(x) * l(y)) / (sqrt(l(x)^2 + l(y)^2) / sqrt(2)) * exp(- (x - y)^2 / (2 * (sqrt(l(x)^2 + l(y)^2) / sqrt(2))^2))
= sqrt(2 * l(x) * l(y) / (l(x)^2 + l(y)^2)) * exp(- (x - y)^2 / (l(x)^2 + l(y)^2))
```
which is exactly the formula in the docstring (and the other issue).